[api-documenter/-extractor/-model] Support 'abstract' Modifier

fwienber opened the pull request:
## Summary

Fixes #3661

Proposal implementation of issue #3661.

Fully supporting the `abstract` modifier in _api-documenter_ involves extending the model (_api-extractor-model_) and letting _api-extractor_ extract the information into the model.

Then, _api-documenter_ can check whether API items have the new ApiAbstractMixin and if the new `isAbstract` property is true, add the modifier to the output. On the fly, the order of modifiers has been slightly adjusted to match the required order in TypeScript syntax.

For abstract classes, instead of adding a "Modifier" column for just this one modifier, a dedicated table has been added, because as an API designer, I think of abstract classes as a third category in between interfaces and (non-abstract) classes. Like interfaces, they cannot be instantiated, but like classes, you can subclass them.

|                | use as type | subclass | instantiate |
-----------------|-------------|----------|-------------|
| interface      | ✅          | ❌       | ❌         |
| abstract class | ✅          | ✅       | ❌         |
| class          | ✅          | ✅       | ✅         |


## Details

There are separate commits for the changes in _api-extractor(-model)_ and _api-documenter_.
Adding a property to the model affects many generated `*.api.json` model files.

The last commit implements the second alternative mentioned in #3661: Instead of adding a "Modifiers" column to the "Classes" overview table, a separate table for "Abstract Classes" is added to the Markdown output.

Note that other output targets beyond Markdown are not yet handled explicitly. Please advise if there is anything left to do.

## How it was tested

Tests for the changes have been complemented in an existing test scenario (_api-extractor_) or added as a new one (_api-documenter_). A new API schema version has been added. (This also updates several `*.api.json` model files.)

A change log entry has been added.

<!-- Have a question?  Ask for help in the chat room: https://rushstack.zulipchat.com/ -->

zelliott commented on apps/api-documenter/src/documenters/MarkdownDocumenter.ts:
```diff
@@ -502,6 +503,11 @@ export class MarkdownDocumenter {
   private _writePackageOrNamespaceTables(output: DocSection, apiContainer: ApiPackage | ApiNamespace): void {
     const configuration: TSDocConfiguration = this._tsdocConfiguration;
 
+    const abstractClassesTable: DocTable = new DocTable({
+      configuration,
+      headerTitles: ['Abstract Class', 'Description']
+    });
```
I'd probably prefer having a "Modifiers" column instead of a separate section for alignment with other API items that have modifiers. What do you think?

fwienber replied:
I wrote that as a side note in the issue:
> ...(my preferred solution) add a dedicated "Abstract Classes" overview table, because in an API, abstract classes are quite something different from a concrete class, sometimes closer to an interface.

As an API designer, I think of abstract classes as a third category in between interfaces and classes. Like interfaces, they cannot be instantiated, but like (non-abstract) classes, you can subclass them.

|                | defines type | subclass | instantiate |
-----------------|--------------|----------|-------------|
| interface      | ✅          | ❌       | ❌         |
| abstract class | ✅          | ✅       | ❌         |
| class          | ✅          | ✅       | ✅         |

zelliott commented on apps/api-documenter/src/documenters/MarkdownDocumenter.ts:
```diff
@@ -1057,6 +1072,14 @@ export class MarkdownDocumenter {
       }
     }
 
+    if (ApiAbstractMixin.isBaseClassOf(apiItem)) {
+      if (apiItem.isAbstract) {
+        section.appendNode(
+          new DocParagraph({ configuration }, [new DocCodeSpan({ configuration, code: 'abstract' })])
+        );
+      }
+    }
```
I'm not sure whether it was intentional or not, but the order of listed modifiers was previously in alphabetical order. This breaks that - maybe we can show `abstract` first above the other modifiers to keep with alphabetical ordering?

fwienber replied:
I didn't realize they were in alphabetical order, and I'm too not sure whether this was coincidence or intentional. If we keep the additional table at all (see above), of course we can keep the alphabetical order. But what would really be nice here would be a way to _configure_ the desired order in api-documenter.

fwienber replied:
I thought about that and have the following suggestion: Wouldn't it make sense to use the modifier order that would be correct in TypeScript code? Unlike other languages like Java, in TypeScript, the order of modifiers is defined strictly, and unfortunately, it is not alphabetical order. In that sense, the current order is already wrong. It seems the correct order is

> (`public` | `protected` | `private`)? (`static`? | `abstract`?) `override`? `readonly`?

(No member can be `static` and `abstract` at the same time, so the order of these two does not matter.)
_api-documenter_ currently only uses `protected`, `static`, `abstract` (new in this PR) and `readonly`, so `public` (default!), `private` (no API!) and `override` (maybe interesting for navigation, but not as a modifier) do not apply.

I'll add a commit changing that and then we can see whether we keep it that way.

fwienber replied:
Btw., in my first comment, I misunderstood you were referring to the order of _member tables_, not to the order of modifiers. So now that I realized what you meant, we still need to discuss whether to add an "Abstact Classes" table at all and if so, where / in what order.

zelliott replied:
> I thought about that and have the following suggestion: Wouldn't it make sense to use the modifier order that would be correct in TypeScript code?

Interesting idea - I think either this or alphabetical both make sense!

zelliott reviewed (commented):
Looks great! Two final comments;

* You'll want to bump the `.api.json` version in `DeserializerContext.ts`. See previous edits to that file for examples.
* It'd be great to have at least one test in `build-tests/` where `isAbstract` is `true`. If we already have one  - great - if not, let's add one.

fwienber commented on libraries/api-extractor-model/src/mixins/ApiAbstractMixin.ts:
```diff
@@ -0,0 +1,113 @@
+ *
+ * @public
+ */
+export function ApiAbstractMixin<TBaseClass extends IApiItemConstructor>(
```
Good point, got to check this!

fwienber replied:
This is somewhat hidden in the grammar, but very clear in this error message:

> TS1242: 'abstract' modifier can only appear on a class, method, or property declaration.

So only the mixin to `ApiProperty` is missing. I'm on it.

fwienber replied:
Please re-review! ;-)

fwienber commented:
Just FYI, I did not forget about this PR, but I'm busy with another project for a couple of days.

fwienber commented:
> * It'd be great to have at least one test in `build-tests/` where `isAbstract` is `true`. If we already have one  - great - if not, let's add one.

There already was an abstract class with an abstract method, now resulting in an `"isAbstract": true` entry in the [expected API JSON](https://github.com/fwienber/rushstack/blob/c4c0d53c79d1e4e90f15756facb3deb8a254d765/build-tests/api-extractor-scenarios/etc/apiItemKinds/api-extractor-scenarios.api.json#L175-L223).
Do you think we need to complement test data for an abstract property?

fwienber commented:
I guess I now addressed all review issues and thus resolved the conversations above. @zelliott, please re-review and of course feel free to re-open the resolved conversation to comment!

fwienber commented:
If you don't like the merge commit that resolves the conflict with "main", I can also rebase my branch, but didn't want to do so while there are comments on the commits that would be lost when force-pushing the rebased branch.

fwienber commented:
I had a bit more time and added more tests:
- added an abstract property to api-extractor-scenarios
- added an abstract class with an abstract method & property to api-documentor-scenarios

Diff in regenerated report files looks good.

fwienber commented:
ping

zelliott commented:
Hoping to take a look this weekend!

zelliott commented:
> There already was an abstract class with an abstract method, now resulting in an "isAbstract": true entry in the [expected API JSON](https://github.com/fwienber/rushstack/blob/c4c0d53c79d1e4e90f15756facb3deb8a254d765/build-tests/api-extractor-scenarios/etc/apiItemKinds/api-extractor-scenarios.api.json#L175-L223).
Do you think we need to complement test data for an abstract property?

No, I think that's good enough then. Thanks for checking!

zelliott reviewed (approved):
Sorry for the very long delay, have been busy with other work (as we all are!). LGTM!

fwienber commented on build-tests/api-documenter-scenarios/src/inheritedMembers/index.ts:
```diff
@@ -124,3 +124,14 @@ export interface IExtendsInterfaceLikeTypeAlias extends IInterfaceLikeTypeAlias,
+  public abstract method(): void;
+  /** Some abstract property. */
+  protected abstract property: number;
+}
```
Okay, I moved the test to api-documenter-test and force-pushed the branch.

fwienber commented:
@zelliott Sorry, this time I missed the notification, then was out of office last week.
I moved the test (last unresolved conversation) as you recommended.
Next step is to find a code owner for review?

zelliott commented:
> @zelliott Sorry, this time I missed the notification, then was out of office last week. I moved the test (last unresolved conversation) as you recommended. Next step is to find a code owner for review?

Yep! LGTM

fwienber commented:
So how is the process here? Are the four mentioned code owners supposed to notice that this PR is ready for review, or do I have to ping (any of) them?

zelliott commented:
I don't think there's any well defined process unfortunately, @octogonz to correct me otherwise. I typically ping @octogonz over Zulip chat or on the PR itself.

fwienber commented:
Okay, magic spell "ping @octogonz" 🧞

octogonz commented:
Sorry I overlooked this PR! 😅 (In the future, you can ping me personally on Zulip if needed.)
